server/src/main/java/org/opensearch/search/aggregations/NonGlobalAggCollectorManagerWithSingleCollector.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.aggregations;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+import org.opensearch.search.internal.SearchContext;
+import org.opensearch.search.profile.query.CollectorResult;
+import org.opensearch.search.query.ReduceableSearchResult;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * {@link CollectorManager} to take care of non-global aggregation operators in case of non-concurrent segment search. This
+ * CollectorManager returns the same collector instance (i.e. created in constructor of super class) on each newCollector call
+ */
+public class NonGlobalAggCollectorManagerWithSingleCollector extends AggregationCollectorManager {
+
+    private final Collector collector;
+
+    public NonGlobalAggCollectorManagerWithSingleCollector(SearchContext context) throws IOException {
+        super(context, context.aggregations().factories()::createTopLevelNonGlobalAggregators, CollectorResult.REASON_AGGREGATION);
+        collector = super.newCollector();
```

Comment:
Minor, we should better check the collector is not null at initialization time, no need to assert in `newCollector()` (and if I remember, Apache Lucene will fail with NPE anyway).

```suggestion
        collector =  Objects.requireNonNull(super.newCollector());
```

Reply:
The assertion is for `NO_OP` collector and non null. IIRC, in the [prototype](https://github.com/reta/OpenSearch/commit/5ddafb5809253c3d46581b733d35bcdfc2136f4f#diff-869e616355ec86b3b7cd6aae1a3a76d8860d873092ca91d5fbfc2ff7edfd718aR37) we were passing the empty `GlobalAggregator` list to `AggregationCollectorManager` which was getting wrapped to `NO_OP` collector. Collector manager for both global and non global aggregation is created only if respective aggregation is present in the context. So in such cases we don't expect the `collector` to be `NO_OP` collector and hence the assertion. This will catch any unexpected cases where `collector` is getting created for aggregation when aggregation is not present in the request.

Reply:
Sorry, my bad, indeed this is NO_OP, (not `null`), should we throw an exception is this case instead of assertion? (assertions could be turned of) Or NO_OP is just undesired but not harmful?

Reply:
IIRC it is undesired and not necessarily harmful

Reply:
So you could probably assert in constructor once, no need to assert on every newCollector call, right?

Reply:
we can but any `newCollector` call should not return the NO_OP collector so I don't see any concern in keeping it there. if there is any bug in `newCollector` that for some reason subsequent call returns NO_OP collector then assert will help to catch in test.

Reply:
@sohami I unresolving this conversation, the the assertion is indeed for `null` and not NO_OP
```
assert collector != null : "Collector instance is null";
```

Reply:
fixed this